### PR TITLE
Laboratory Work #1, Skachkova Anastasia - 20FPL3

### PR DESCRIPTION
**Pull request by awakenedlilac**

### CODE REVIEW

File: scrapper.py [hunk 1]
```diff
@@ -43,21 +56,64 @@ def get_search_urls(self):
...
     Creates ASSETS_PATH folder if not created and removes existing folder
     """
-    pass
+    try:
+        os.rmdir(ASSETS_PATH)
+    except FileNotFoundError:
+        os.mkdir(ASSETS_PATH)
 
 
 def validate_config(crawler_path):
```
Comment (marina-kaz) [inline scrapper.py:93]: what about other cases?

**Review by marina-kaz — CHANGES_REQUESTED** [review]
It is time to implement core logic of this work :)

File: scrapper.py [hunk 2]
```diff
@@ -1,7 +1,13 @@
...
 Scrapper implementation
 """
-
+import json
+import re
+import requests
+import shutil
+from bs4 import BeautifulSoup
+from constants import ASSETS_PATH, CRAWLER_CONFIG_PATH
+from core_utils.article import Article
```
Comment (marina-kaz) [inline scrapper.py:10]: it is better to separate imports of external and internal modules
moreover, they should be in alphabetical order

File: scrapper.py [hunk 3]
```diff
@@ -20,44 +26,101 @@ class IncorrectNumberOfArticlesError(Exception):
...
     """
     def __init__(self, seed_urls, max_articles: int):
-        pass
+        self.seed_urls = seed_urls
+        self.max_articles = max_articles
+        self.urls = []
 
     def _extract_url(self, article_bs):
-        pass
+        self.urls.append('https://gazeta.ru' + article_bs.find('a')['href'])
```
Comment (marina-kaz) [inline scrapper.py:39]: do you only extract one url from each of your seed urls?

File: scrapper.py [hunk 4]
```diff
@@ -20,44 +26,101 @@ class IncorrectNumberOfArticlesError(Exception):
...
     """
     Validates given config
     """
-    pass
+    with open(crawler_path, 'r', encoding='utf-8') as file:
+        config = json.load(file)
+    seed_urls = config['seed_urls']
+    max_articles = config['total_articles_to_find_and_parse']
+    if not isinstance(max_articles, int) or max_articles <= 0:
+        raise IncorrectNumberOfArticlesError
```
Comment (marina-kaz) [inline scrapper.py:104]: in case max_articles is in fact not an integer, this line will fail
you should separate these checks

**Review by marina-kaz — CHANGES_REQUESTED** [review]
You are getting there, keep going

File: scrapper.py [hunk 5]
```diff
@@ -20,44 +30,129 @@ class IncorrectNumberOfArticlesError(Exception):
...
         """
-        pass
+        for url in self.seed_urls:
+            sleep(randint(1, 5))
+            response = requests.get(url)
+            page_bs = BeautifulSoup(response.text, 'lxml')
+            urls = self._extract_url(page_bs)
+            for the_url in urls:
+                if len(self.urls) < self.max_articles:
+                    self.urls.append(the_url)
```
Comment (marina-kaz) [inline scrapper.py:68]: before adding a url, check that it has not yet been added

File: scrapper.py [hunk 6]
```diff
@@ -20,44 +30,129 @@ class IncorrectNumberOfArticlesError(Exception):
...
+        config = json.load(file)
+    seed_urls = config['seed_urls']
+    max_articles = config['total_articles_to_find_and_parse']
+    if not isinstance(max_articles, int):
+        raise IncorrectNumberOfArticlesError
+    if max_articles <= 0:
+        raise IncorrectNumberOfArticlesError
+    if not isinstance(seed_urls, list) or not seed_urls:
+        raise IncorrectURLError
+    if max_articles > 100:
```
Comment (marina-kaz) [inline scrapper.py:139]: take a bigger number, this is too close to how many articles you are required to collect

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Just a little more to fix

File: scrapper.py (hunk 2 above)
Comment (awakenedlilac) [reply scrapper.py:11]: when i put them alphabetically, code style says me to put them the way i did.

Comment (marina-kaz) [reply scrapper.py:11]: code style tells you to place standard imports before external and external before local, and it is correct
What is also correct is that to make code more readable it is better to keep alphabetical order among standard imports, among external and among local
It does not contradict

File: config/stage_2_crawler_tests/s2_5_raw_metadata_test.py [hunk 7]
```diff
@@ -73,6 +73,7 @@ def test_validate_sort_metadata(self):
     @pytest.mark.mark8
     @pytest.mark.mark10
     @pytest.mark.stage_2_5_dataset_validation
+    @pytest.mark.skip(reason='because of nbsp spaces')
```
Comment (marina-kaz) [inline config/stage_2_crawler_tests/s2_5_raw_metadata_test.py:76]: remind me, please: who have you discussed the necessity of test skip with?

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Time to finish Crawler and embark on Pipeline

File: config/stage_2_crawler_tests/s2_5_raw_metadata_test.py (hunk 7 above)
Comment (awakenedlilac) [reply config/stage_2_crawler_tests/s2_5_raw_metadata_test.py:76]: we have discussed it on the lecture: Alexander Vladimirovich told me to skip this test, due to nbsp spaces on my source.

Comment (marina-kaz) [reply config/stage_2_crawler_tests/s2_5_raw_metadata_test.py:76]: OK, thank you

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Good job!

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Time to speed up with the Pipeline :)

**Review by marina-kaz — CHANGES_REQUESTED** [review]
No significant changes to review 
Deadline is approaching

File: pipeline.py [hunk 8]
```diff
@@ -72,31 +91,96 @@ class TextProcessingPipeline:
...
+            tokens_for_article = []
+            single_tagged_tokens = []
+            multiple_tagged_tokens = []
+            for token in tokens:
+                tokens_for_article.append(token.get_cleaned())
+                single_tagged_tokens.append(token.get_single_tagged())
+                multiple_tagged_tokens.append(token.get_multiple_tagged())
+            article.save_as(' '.join(tokens_for_article), kind='cleaned')
+            article.save_as(' '.join(single_tagged_tokens), kind='single_tagged')
+            article.save_as(' '.join(multiple_tagged_tokens), kind='multiple_tagged')
```
Comment (marina-kaz) [inline pipeline.py]: as it is stated in the 5.3 section of pipeline documentation, you are required to pass `ArtifactType` attributes as a `kind` argument to the `save_as() method`, this way code is more secure as it is protected from almost untraceable errors caused by simple typos

File: pipeline.py [hunk 9]
```diff
@@ -72,31 +91,96 @@ class TextProcessingPipeline:
...
+    if not path.is_dir():
+        raise NotADirectoryError
+    all_article_ids = []
+    for file in path.glob("*.txt"):
+        with open(file, 'r', encoding='utf-8') as text_file:
+            text = text_file.read()
+        raws = list(path.glob('*_raw.txt'))
+        metas = list(path.glob('*_meta.json'))
+        if not len(metas) == len(raws):
+            raise InconsistentDatasetError
```
Comment (marina-kaz) [inline pipeline.py]: I do not understand why you have to verify equality of numbers of meta and raw files within this cycle (lines 153-156)

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Good job!

**Pull request closed without merging**